<!DOCTYPE html>
<html lang="zh-hans" prefix="og: http://ogp.me/ns# fb: https://www.facebook.com/2008/fbml" class="han-init">
<head>
    <title>asyncio 学习笔记：并发执行 Task - Huang Huang 的博客</title>
    <!-- Using the latest rendering mode for IE -->
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- share.css -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/social-share.js/1.0.16/css/share.min.css">
    <!-- share.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/social-share.js/1.0.16/js/social-share.min.js"></script>


    <link href="https://mozillazg.com/favicon.ico" rel="icon">

<link rel="canonical" href="https://mozillazg.com/2017/08/python-asyncio-note-task-usage.html">

        <meta name="author" content="mozillazg" />
        <meta name="keywords" content="asyncio" />
    <meta name="description" content="本文是 https://pymotw.com/3/asyncio/tasks.html 的学习笔记，强烈推荐直接阅读原文。 Task 是跟事件循环交互的一种主要方式。Task 包装并追踪 Coroutine 的完成状态。Task 是 Future 的子类，因此其他 Coroutine 可以 wait Task 并且在 Task 完成时还可以获取 Coroutine 的结果。 启动 Task¶ 可以使用 create_task 来创建一个 Task 实例: # asyncio_create_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;waiting for {!r}&#39;.format(task)) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 可以看到 await task 返回的是 task_func 函数的返回值: $ python3.6 asyncio_create_task.py creating task waiting for &lt;Task pending coro=&lt;task_func() running at asyncio_create_task.py:5&gt;&gt; in task_func task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_create_task.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 取消 Task¶ 可以在 Task 完成之前取消 task 的操作: # asyncio_cancel_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;canceling task&#39;) task.cancel() print(&#39;canceled task {!r}&#39;.format(task)) try: await task except asyncio.CancelledError: print(&#39;caught error from canceled task&#39;) else: print(&#39;task result: {!r}&#39;.format(task.result())) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 当在事件循环启动前取消一下 task 时， await task 会抛出 CancelledError 异常: $ python3.6 asyncio_cancel_task.py creating task canceling task canceled task &lt;Task cancelling coro=&lt;task_func() running at asyncio_cancel_task.py:5&gt;&gt; caught error from canceled task 当取消一个正在等待其他 concurrent 操作的 task 时，等待操作的位置会抛出 CancelledError 异常: # asyncio_cancel_task2.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) raise return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) t.cancel() print(&#39;canceled the task&#39;) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) try: await task except asyncio.CancelledError: print(&#39;main() also sees task as canceled&#39;) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task2.py creating task in task_func, sleeping in task_canceller canceled the task task_func was canceled main() also sees task as canceled 上面 asyncio_cancel_task2.py 的 task_func 中如果没有把 CancelledError 异常再 raise 出来的话，其实还是会继续执行下去的: # asyncio_cancel_task3.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) print(&#39;task_func still active&#39;) return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) ok = t.cancel() print(&#39;canceled the task: {0}&#39;.format(ok)) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task3.py creating task in task_func, sleeping in task_canceller canceled the task: True task_func was canceled task_func still active task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_cancel_task3.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 在 Coroutine 中创建 Task¶ ensure_future() 函数返回一个与一个 coroutine 的执行相关连的 Task。这个 Task 实例可以作为变量传入其他代码中，这样的话其他代码就可以直接 await 这个 Task 而不需要知道原始的 coroutine 是如何被创建的。 # asyncio_ensure_future.py import asyncio async def wrapped(): print(&#39;wrapped&#39;) return &#39;result&#39; async def inner(task): print(&#39;inner: starting&#39;) print(&#39;inner: waiting for {!r}&#39;.format(task)) result = await task print(&#39;inner: task returned {!r}&#39;.format(result)) async def starter(): print(&#39;starter: creating task&#39;) task = asyncio.ensure_future(wrapped()) print(&#39;starter: waiting for inner&#39;) await inner(task) print(&#39;starter: inner returned&#39;) event_loop = asyncio.get_event_loop() try: print(&#39;entering event loop&#39;) result = event_loop.run_until_complete(starter()) finally: event_loop.close() 需要注意的是，传入 ensure_future() 的 coroutine 不会立马启动，需要有某个地方使用了 await 语句操作创建的 task 的时候它才会被执行。 $ python3.6 asyncio_ensure_future.py entering event loop starter: creating task starter: waiting for inner inner: starting inner: waiting for &lt;Task pending coro=&lt;wrapped() running at asyncio_ensure_future.py:5&gt;&gt; wrapped inner: task returned &#39;result&#39; starter: inner returned 参考资料¶ Executing Tasks Concurrently — PyMOTW 3 18.5.3. Tasks and coroutines — Python 3.6.2 documentation" />

    <style>
      .js-toc {
        margin-bottom: 20px;
      }
      .donate-modal {
        text-align: center;
      }
      #donate-modal-container .donate-image {
        max-height: 300px !important;
        min-height: inherit !important;
      }
    </style>

        <meta property="og:site_name" content="Huang Huang 的博客" />
        <meta property="og:type" content="article"/>
        <meta property="og:title" content="asyncio 学习笔记：并发执行 Task"/>
        <meta property="og:url" content="https://mozillazg.com/2017/08/python-asyncio-note-task-usage.html"/>
        <meta property="og:description" content="本文是 https://pymotw.com/3/asyncio/tasks.html 的学习笔记，强烈推荐直接阅读原文。 Task 是跟事件循环交互的一种主要方式。Task 包装并追踪 Coroutine 的完成状态。Task 是 Future 的子类，因此其他 Coroutine 可以 wait Task 并且在 Task 完成时还可以获取 Coroutine 的结果。 启动 Task¶ 可以使用 create_task 来创建一个 Task 实例: # asyncio_create_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;waiting for {!r}&#39;.format(task)) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 可以看到 await task 返回的是 task_func 函数的返回值: $ python3.6 asyncio_create_task.py creating task waiting for &lt;Task pending coro=&lt;task_func() running at asyncio_create_task.py:5&gt;&gt; in task_func task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_create_task.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 取消 Task¶ 可以在 Task 完成之前取消 task 的操作: # asyncio_cancel_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;canceling task&#39;) task.cancel() print(&#39;canceled task {!r}&#39;.format(task)) try: await task except asyncio.CancelledError: print(&#39;caught error from canceled task&#39;) else: print(&#39;task result: {!r}&#39;.format(task.result())) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 当在事件循环启动前取消一下 task 时， await task 会抛出 CancelledError 异常: $ python3.6 asyncio_cancel_task.py creating task canceling task canceled task &lt;Task cancelling coro=&lt;task_func() running at asyncio_cancel_task.py:5&gt;&gt; caught error from canceled task 当取消一个正在等待其他 concurrent 操作的 task 时，等待操作的位置会抛出 CancelledError 异常: # asyncio_cancel_task2.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) raise return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) t.cancel() print(&#39;canceled the task&#39;) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) try: await task except asyncio.CancelledError: print(&#39;main() also sees task as canceled&#39;) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task2.py creating task in task_func, sleeping in task_canceller canceled the task task_func was canceled main() also sees task as canceled 上面 asyncio_cancel_task2.py 的 task_func 中如果没有把 CancelledError 异常再 raise 出来的话，其实还是会继续执行下去的: # asyncio_cancel_task3.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) print(&#39;task_func still active&#39;) return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) ok = t.cancel() print(&#39;canceled the task: {0}&#39;.format(ok)) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task3.py creating task in task_func, sleeping in task_canceller canceled the task: True task_func was canceled task_func still active task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_cancel_task3.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 在 Coroutine 中创建 Task¶ ensure_future() 函数返回一个与一个 coroutine 的执行相关连的 Task。这个 Task 实例可以作为变量传入其他代码中，这样的话其他代码就可以直接 await 这个 Task 而不需要知道原始的 coroutine 是如何被创建的。 # asyncio_ensure_future.py import asyncio async def wrapped(): print(&#39;wrapped&#39;) return &#39;result&#39; async def inner(task): print(&#39;inner: starting&#39;) print(&#39;inner: waiting for {!r}&#39;.format(task)) result = await task print(&#39;inner: task returned {!r}&#39;.format(result)) async def starter(): print(&#39;starter: creating task&#39;) task = asyncio.ensure_future(wrapped()) print(&#39;starter: waiting for inner&#39;) await inner(task) print(&#39;starter: inner returned&#39;) event_loop = asyncio.get_event_loop() try: print(&#39;entering event loop&#39;) result = event_loop.run_until_complete(starter()) finally: event_loop.close() 需要注意的是，传入 ensure_future() 的 coroutine 不会立马启动，需要有某个地方使用了 await 语句操作创建的 task 的时候它才会被执行。 $ python3.6 asyncio_ensure_future.py entering event loop starter: creating task starter: waiting for inner inner: starting inner: waiting for &lt;Task pending coro=&lt;wrapped() running at asyncio_ensure_future.py:5&gt;&gt; wrapped inner: task returned &#39;result&#39; starter: inner returned 参考资料¶ Executing Tasks Concurrently — PyMOTW 3 18.5.3. Tasks and coroutines — Python 3.6.2 documentation"/>
        <meta property="article:published_time" content="2017-08-20" />
            <meta property="article:section" content="python" />
            <meta property="article:tag" content="asyncio" />
            <meta property="article:author" content="mozillazg" />
            <meta property="og:image"
                  content="https://mozillazg.com/static/avatar.jpeg"/>


    <meta name="twitter:card" content="summary">
        <meta name="twitter:site" content="@mozillazg">
        <meta name="twitter:creator" content="@mozillazg">
    <meta name="twitter:domain" content="https://mozillazg.com">
            <meta property="twitter:image"
                  content="https://mozillazg.com/static/avatar.jpeg"/>

    <!-- Bootstrap -->
        <link rel="stylesheet" href="https://mozillazg.com/theme/css/bootstrap.min.css" type="text/css"/>
    <link href="https://mozillazg.com/theme/css/font-awesome.min.css" rel="stylesheet">

    <link href="https://mozillazg.com/theme/css/pygments/github.css" rel="stylesheet">
    <link rel="stylesheet" href="https://mozillazg.com/theme/css/style.css" type="text/css"/>
            <link href="https://mozillazg.com/static/custom.css" rel="stylesheet">

        <link href="https://mozillazg.com/feeds/all.atom.xml" type="application/atom+xml" rel="alternate"
              title="Huang Huang 的博客 ATOM Feed"/>

        <link href="https://mozillazg.com/feeds/all.rss.xml" type="application/rss+xml" rel="alternate"
              title="Huang Huang 的博客 RSS Feed"/>


        <link href="https://mozillazg.com/feeds/python.atom.xml" type="application/atom+xml" rel="alternate"
              title="Huang Huang 的博客 python ATOM Feed"/>


    <script type="application/ld+json">
    {
        "@context": "https://schema.org",
        "@type": "Website",
        "publisher": {
            "@type": "Person",
            "name": "mozillazg",
            "logo": "https://mozillazg.com/static/avatar.jpeg"
        },
        "author": {
            "@type": "Person",
            "name": "mozillazg",
            "image": "https://mozillazg.com/static/avatar.jpeg",
            "url": "https://mozillazg.com",
            "sameAs": []
        },
        // "headline": "",
        "url": "https://mozillazg.com/2017/08/python-asyncio-note-task-usage.html",
        "datePublished": "2017-08-20",
        // "dateModified": "",
        "image": "https://mozillazg.com/static/avatar.jpeg",
        "keywords": "asyncio",
        "description": "本文是 https://pymotw.com/3/asyncio/tasks.html 的学习笔记，强烈推荐直接阅读原文。 Task 是跟事件循环交互的一种主要方式。Task 包装并追踪 Coroutine 的完成状态。Task 是 Future 的子类，因此其他 Coroutine 可以 wait Task 并且在 Task 完成时还可以获取 Coroutine 的结果。 启动 Task¶ 可以使用 create_task 来创建一个 Task 实例: # asyncio_create_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;waiting for {!r}&#39;.format(task)) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 可以看到 await task 返回的是 task_func 函数的返回值: $ python3.6 asyncio_create_task.py creating task waiting for &lt;Task pending coro=&lt;task_func() running at asyncio_create_task.py:5&gt;&gt; in task_func task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_create_task.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 取消 Task¶ 可以在 Task 完成之前取消 task 的操作: # asyncio_cancel_task.py import asyncio async def task_func(): print(&#39;in task_func&#39;) return &#39;the result&#39; async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) print(&#39;canceling task&#39;) task.cancel() print(&#39;canceled task {!r}&#39;.format(task)) try: await task except asyncio.CancelledError: print(&#39;caught error from canceled task&#39;) else: print(&#39;task result: {!r}&#39;.format(task.result())) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 当在事件循环启动前取消一下 task 时， await task 会抛出 CancelledError 异常: $ python3.6 asyncio_cancel_task.py creating task canceling task canceled task &lt;Task cancelling coro=&lt;task_func() running at asyncio_cancel_task.py:5&gt;&gt; caught error from canceled task 当取消一个正在等待其他 concurrent 操作的 task 时，等待操作的位置会抛出 CancelledError 异常: # asyncio_cancel_task2.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) raise return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) t.cancel() print(&#39;canceled the task&#39;) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) try: await task except asyncio.CancelledError: print(&#39;main() also sees task as canceled&#39;) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task2.py creating task in task_func, sleeping in task_canceller canceled the task task_func was canceled main() also sees task as canceled 上面 asyncio_cancel_task2.py 的 task_func 中如果没有把 CancelledError 异常再 raise 出来的话，其实还是会继续执行下去的: # asyncio_cancel_task3.py import asyncio async def task_func(): print(&#39;in task_func, sleeping&#39;) try: await asyncio.sleep(1) except asyncio.CancelledError: print(&#39;task_func was canceled&#39;) print(&#39;task_func still active&#39;) return &#39;the result&#39; def task_canceller(t): print(&#39;in task_canceller&#39;) ok = t.cancel() print(&#39;canceled the task: {0}&#39;.format(ok)) async def main(loop): print(&#39;creating task&#39;) task = loop.create_task(task_func()) loop.call_soon(task_canceller, task) return_value = await task print(&#39;task completed {!r}&#39;.format(task)) print(&#39;return value: {!r}&#39;.format(return_value)) event_loop = asyncio.get_event_loop() try: event_loop.run_until_complete(main(event_loop)) finally: event_loop.close() 结果: $ python3.6 asyncio_cancel_task3.py creating task in task_func, sleeping in task_canceller canceled the task: True task_func was canceled task_func still active task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_cancel_task3.py:5&gt; result=&#39;the result&#39;&gt; return value: &#39;the result&#39; 在 Coroutine 中创建 Task¶ ensure_future() 函数返回一个与一个 coroutine 的执行相关连的 Task。这个 Task 实例可以作为变量传入其他代码中，这样的话其他代码就可以直接 await 这个 Task 而不需要知道原始的 coroutine 是如何被创建的。 # asyncio_ensure_future.py import asyncio async def wrapped(): print(&#39;wrapped&#39;) return &#39;result&#39; async def inner(task): print(&#39;inner: starting&#39;) print(&#39;inner: waiting for {!r}&#39;.format(task)) result = await task print(&#39;inner: task returned {!r}&#39;.format(result)) async def starter(): print(&#39;starter: creating task&#39;) task = asyncio.ensure_future(wrapped()) print(&#39;starter: waiting for inner&#39;) await inner(task) print(&#39;starter: inner returned&#39;) event_loop = asyncio.get_event_loop() try: print(&#39;entering event loop&#39;) result = event_loop.run_until_complete(starter()) finally: event_loop.close() 需要注意的是，传入 ensure_future() 的 coroutine 不会立马启动，需要有某个地方使用了 await 语句操作创建的 task 的时候它才会被执行。 $ python3.6 asyncio_ensure_future.py entering event loop starter: creating task starter: waiting for inner inner: starting inner: waiting for &lt;Task pending coro=&lt;wrapped() running at asyncio_ensure_future.py:5&gt;&gt; wrapped inner: task returned &#39;result&#39; starter: inner returned 参考资料¶ Executing Tasks Concurrently — PyMOTW 3 18.5.3. Tasks and coroutines — Python 3.6.2 documentation",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https://mozillazg.com"
        }
    }
    </script>

</head>
<body>

<div class="navbar" role="navigation">
	<div class="container">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-ex1-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a href="https://mozillazg.com/" class="navbar-brand">
Huang Huang 的博客            </a>
        </div>
        <div class="collapse navbar-collapse navbar-ex1-collapse">
            <ul class="nav navbar-nav">
                    <li><a href="https://mozillazg.com/feeds/all.atom.xml">Feed</a></li>
                    <li><a href="/2014/10/pages/about-me.html">About</a></li>
            </ul>
            <ul class="nav navbar-nav navbar-right">
              <li><a href="https://mozillazg.com/archives.html"><i class="fa fa-th-list"></i><span class="icon-label">Archives</span></a></li>
            </ul>
        </div>
        <!-- /.navbar-collapse -->
    </div>
</div> <!-- /.navbar -->
<!-- Banner -->
<!-- End Banner -->
<div class="container">
    <div class="row">
        <div class="col-lg-12">
    <section id="content" class="yue">
        <article>
            <header class="text-center">
                <h1>
                    <a href="https://mozillazg.com/2017/08/python-asyncio-note-task-usage.html"
                       rel="bookmark"
                       title="Permalink to asyncio 学习笔记：并发执行 Task">
                        asyncio 学习笔记：并发执行 Task
                    </a>
                </h1>
                <p class="published">
                    <time datetime="2017-08-20T00:00:00+00:00">
                    2017-08-20
                    </time>
                </p>
            </header>
            <div class="entry-content">
                <div class="well-sm article-info">
<footer class="post-info">
        <a href="https://mozillazg.com/category/python.html">python</a>


<span class="label label-default hide">Tags</span>
	<a href="https://mozillazg.com/tag/asyncio.html">asyncio</a>
    
</footer><!-- /.post-info -->                </div>
                <div class="js-toc"></div>
                <div>
                <p>本文是 <a class="reference external" href="https://pymotw.com/3/asyncio/tasks.html">https://pymotw.com/3/asyncio/tasks.html</a>
的学习笔记，强烈推荐直接阅读原文。</p>
<p>Task 是跟事件循环交互的一种主要方式。Task 包装并追踪 Coroutine
的完成状态。Task 是 <tt class="docutils literal">Future</tt> 的子类，因此其他 Coroutine 可以 wait Task
并且在 Task 完成时还可以获取 Coroutine 的结果。</p>
<div class="section" id="task">
<h2 id="hidtask">启动 Task<a class="headerlink" href="#hidtask" title="Permalink to this headline">¶</a></h2>
<p>可以使用 <tt class="docutils literal">create_task</tt> 来创建一个 <tt class="docutils literal">Task</tt> 实例:</p>
<pre class="code python literal-block">
<span class="c1"># asyncio_create_task.py</span>
<span class="kn">import</span> <span class="nn">asyncio</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">task_func</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_func'</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">'the result'</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">loop</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'creating task'</span><span class="p">)</span>
    <span class="n">task</span> <span class="o">=</span> <span class="n">loop</span><span class="o">.</span><span class="n">create_task</span><span class="p">(</span><span class="n">task_func</span><span class="p">())</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'waiting for {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="p">))</span>
    <span class="n">return_value</span> <span class="o">=</span> <span class="n">await</span> <span class="n">task</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'task completed {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="p">))</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'return value: {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">return_value</span><span class="p">))</span>


<span class="n">event_loop</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">get_event_loop</span><span class="p">()</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">run_until_complete</span><span class="p">(</span><span class="n">main</span><span class="p">(</span><span class="n">event_loop</span><span class="p">))</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre>
<p>可以看到 <tt class="docutils literal">await task</tt> 返回的是 <tt class="docutils literal">task_func</tt> 函数的返回值:</p>
<pre class="literal-block">
$ python3.6 asyncio_create_task.py
creating task
waiting for &lt;Task pending coro=&lt;task_func() running at asyncio_create_task.py:5&gt;&gt;
in task_func
task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_create_task.py:5&gt; result='the result'&gt;
return value: 'the result'
</pre>
</div>
<div class="section" id="id1">
<h2 id="hidid1">取消 Task<a class="headerlink" href="#hidid1" title="Permalink to this headline">¶</a></h2>
<p>可以在 <tt class="docutils literal">Task</tt> 完成之前取消 task 的操作:</p>
<pre class="code python literal-block">
<span class="c1"># asyncio_cancel_task.py</span>
<span class="kn">import</span> <span class="nn">asyncio</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">task_func</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_func'</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">'the result'</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">loop</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'creating task'</span><span class="p">)</span>
    <span class="n">task</span> <span class="o">=</span> <span class="n">loop</span><span class="o">.</span><span class="n">create_task</span><span class="p">(</span><span class="n">task_func</span><span class="p">())</span>

    <span class="k">print</span><span class="p">(</span><span class="s1">'canceling task'</span><span class="p">)</span>
    <span class="n">task</span><span class="o">.</span><span class="n">cancel</span><span class="p">()</span>

    <span class="k">print</span><span class="p">(</span><span class="s1">'canceled task {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="p">))</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">await</span> <span class="n">task</span>
    <span class="k">except</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">CancelledError</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s1">'caught error from canceled task'</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s1">'task result: {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="o">.</span><span class="n">result</span><span class="p">()))</span>


<span class="n">event_loop</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">get_event_loop</span><span class="p">()</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">run_until_complete</span><span class="p">(</span><span class="n">main</span><span class="p">(</span><span class="n">event_loop</span><span class="p">))</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre>
<p>当在事件循环启动前取消一下 task 时， <tt class="docutils literal">await task</tt> 会抛出
<tt class="docutils literal">CancelledError</tt> 异常:</p>
<pre class="literal-block">
$ python3.6 asyncio_cancel_task.py
creating task
canceling task
canceled task &lt;Task cancelling coro=&lt;task_func() running at asyncio_cancel_task.py:5&gt;&gt;
caught error from canceled task
</pre>
<p>当取消一个正在等待其他 concurrent 操作的 task 时，等待操作的位置会抛出
<tt class="docutils literal">CancelledError</tt> 异常:</p>
<pre class="code python literal-block">
<span class="c1"># asyncio_cancel_task2.py</span>
<span class="kn">import</span> <span class="nn">asyncio</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">task_func</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_func, sleeping'</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">await</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">CancelledError</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s1">'task_func was canceled'</span><span class="p">)</span>
        <span class="k">raise</span>
    <span class="k">return</span> <span class="s1">'the result'</span>


<span class="k">def</span> <span class="nf">task_canceller</span><span class="p">(</span><span class="n">t</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_canceller'</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">cancel</span><span class="p">()</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'canceled the task'</span><span class="p">)</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">loop</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'creating task'</span><span class="p">)</span>
    <span class="n">task</span> <span class="o">=</span> <span class="n">loop</span><span class="o">.</span><span class="n">create_task</span><span class="p">(</span><span class="n">task_func</span><span class="p">())</span>
    <span class="n">loop</span><span class="o">.</span><span class="n">call_soon</span><span class="p">(</span><span class="n">task_canceller</span><span class="p">,</span> <span class="n">task</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">await</span> <span class="n">task</span>
    <span class="k">except</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">CancelledError</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s1">'main() also sees task as canceled'</span><span class="p">)</span>


<span class="n">event_loop</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">get_event_loop</span><span class="p">()</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">run_until_complete</span><span class="p">(</span><span class="n">main</span><span class="p">(</span><span class="n">event_loop</span><span class="p">))</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre>
<p>结果:</p>
<pre class="literal-block">
$ python3.6 asyncio_cancel_task2.py
creating task
in task_func, sleeping
in task_canceller
canceled the task
task_func was canceled
main() also sees task as canceled
</pre>
<p>上面 asyncio_cancel_task2.py 的 <tt class="docutils literal">task_func</tt> 中如果没有把
<tt class="docutils literal">CancelledError</tt> 异常再 raise 出来的话，其实还是会继续执行下去的:</p>
<pre class="code python literal-block">
<span class="c1"># asyncio_cancel_task3.py</span>
<span class="kn">import</span> <span class="nn">asyncio</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">task_func</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_func, sleeping'</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">await</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">CancelledError</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s1">'task_func was canceled'</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'task_func still active'</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">'the result'</span>


<span class="k">def</span> <span class="nf">task_canceller</span><span class="p">(</span><span class="n">t</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'in task_canceller'</span><span class="p">)</span>
    <span class="n">ok</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">cancel</span><span class="p">()</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'canceled the task: {0}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ok</span><span class="p">))</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">loop</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'creating task'</span><span class="p">)</span>
    <span class="n">task</span> <span class="o">=</span> <span class="n">loop</span><span class="o">.</span><span class="n">create_task</span><span class="p">(</span><span class="n">task_func</span><span class="p">())</span>
    <span class="n">loop</span><span class="o">.</span><span class="n">call_soon</span><span class="p">(</span><span class="n">task_canceller</span><span class="p">,</span> <span class="n">task</span><span class="p">)</span>
    <span class="n">return_value</span> <span class="o">=</span> <span class="n">await</span> <span class="n">task</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'task completed {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="p">))</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'return value: {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">return_value</span><span class="p">))</span>


<span class="n">event_loop</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">get_event_loop</span><span class="p">()</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">run_until_complete</span><span class="p">(</span><span class="n">main</span><span class="p">(</span><span class="n">event_loop</span><span class="p">))</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre>
<p>结果:</p>
<pre class="literal-block">
$ python3.6 asyncio_cancel_task3.py
creating task
in task_func, sleeping
in task_canceller
canceled the task: True
task_func was canceled
task_func still active
task completed &lt;Task finished coro=&lt;task_func() done, defined at asyncio_cancel_task3.py:5&gt; result='the result'&gt;
return value: 'the result'
</pre>
</div>
<div class="section" id="coroutine-task">
<h2 id="hidcoroutine-task">在 Coroutine 中创建 Task<a class="headerlink" href="#hidcoroutine-task" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal">ensure_future()</tt> 函数返回一个与一个 coroutine 的执行相关连的
Task。这个 Task
实例可以作为变量传入其他代码中，这样的话其他代码就可以直接 await 这个
Task 而不需要知道原始的 coroutine 是如何被创建的。</p>
<pre class="code python literal-block">
<span class="c1"># asyncio_ensure_future.py</span>
<span class="kn">import</span> <span class="nn">asyncio</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">wrapped</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'wrapped'</span><span class="p">)</span>
    <span class="k">return</span> <span class="s1">'result'</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">inner</span><span class="p">(</span><span class="n">task</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'inner: starting'</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'inner: waiting for {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">task</span><span class="p">))</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">await</span> <span class="n">task</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'inner: task returned {!r}'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">result</span><span class="p">))</span>


<span class="n">async</span> <span class="k">def</span> <span class="nf">starter</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'starter: creating task'</span><span class="p">)</span>
    <span class="n">task</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">ensure_future</span><span class="p">(</span><span class="n">wrapped</span><span class="p">())</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'starter: waiting for inner'</span><span class="p">)</span>
    <span class="n">await</span> <span class="n">inner</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'starter: inner returned'</span><span class="p">)</span>


<span class="n">event_loop</span> <span class="o">=</span> <span class="n">asyncio</span><span class="o">.</span><span class="n">get_event_loop</span><span class="p">()</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'entering event loop'</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">event_loop</span><span class="o">.</span><span class="n">run_until_complete</span><span class="p">(</span><span class="n">starter</span><span class="p">())</span>
<span class="k">finally</span><span class="p">:</span>
    <span class="n">event_loop</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre>
<p>需要注意的是，传入 <tt class="docutils literal">ensure_future()</tt> 的 coroutine
不会立马启动，需要有某个地方使用了 <tt class="docutils literal">await</tt> 语句操作创建的 task
的时候它才会被执行。</p>
<pre class="literal-block">
$ python3.6 asyncio_ensure_future.py
entering event loop
starter: creating task
starter: waiting for inner
inner: starting
inner: waiting for &lt;Task pending coro=&lt;wrapped() running at asyncio_ensure_future.py:5&gt;&gt;
wrapped
inner: task returned 'result'
starter: inner returned
</pre>
</div>
<div class="section" id="id2">
<h2 id="hidid2">参考资料<a class="headerlink" href="#hidid2" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="https://pymotw.com/3/asyncio/tasks.html">Executing Tasks Concurrently — PyMOTW
3</a></li>
<li><a class="reference external" href="https://docs.python.org/3/library/asyncio-task.html">18.5.3. Tasks and coroutines — Python 3.6.2
documentation</a></li>
</ul>
</div>

                </div>
            </div>
            <!-- /.entry-content -->
<section class="text-center">
  
<div id="donate"></div>

<div class="social-share"></div>
<div>
<p>有任何意见欢迎在下方留言或者加我微信交流</p>
<img src="/static/wechat.png" height="200px" width="200px"/>
</div>

</section>
<section class="well" id="related-posts">
    <p>Related Posts:</p>
    <ul>
        <li><a href="https://mozillazg.com/2017/08/python-asyncio-note-io-protocol.html">asyncio 学习笔记：使用抽象类 Protocol 实现异步 I/O</a></li>
        <li><a href="https://mozillazg.com/2017/08/python-asyncio-note-coroutines-base-usage.html">asyncio 学习笔记：基本用法</a></li>
        <li><a href="https://mozillazg.com/2017/08/python-asyncio-note-control-Coroutines.html">asyncio 学习笔记：控制组合式 Coroutines</a></li>
        <li><a href="https://mozillazg.com/2017/08/python-asyncio-note-synchronization-primitives.html">asyncio 学习笔记：同步原语</a></li>
        <li><a href="https://mozillazg.com/2017/08/python-asyncio-note-coroutine-schedule.html">asyncio 学习笔记：调用定时函数</a></li>
    </ul>
</section>
    <hr/>
    <section class="comments" id="comments">
        <h2>Comments</h2>

        <div id="disqus_thread"></div>
        <script type="text/javascript">
            /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
            var disqus_shortname = 'my-github-blog'; // required: replace example with your forum shortname

                    var disqus_identifier = 'python-asyncio-note-task-usage';
                var disqus_url = 'https://mozillazg.com/2017/08/python-asyncio-note-task-usage.html';

            var disqus_config = function () {
                this.language = "zh-hans";
            };

            /* * * DON'T EDIT BELOW THIS LINE * * */
            (function () {
                var dsq = document.createElement('script');
                dsq.type = 'text/javascript';
                dsq.async = true;
                dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
        </script>
        <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by
            Disqus.</a></noscript>
        <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

    </section>
        </article>
    </section>

        </div>
    </div>
</div>
<footer>
   <div class="container">
      <hr>
      <div class="row">
         <div class="col-xs-10">&copy; 2019 mozillazg
            &middot; Powered by <a href="https://github.com/DandyDev/pelican-bootstrap3" target="_blank">pelican-bootstrap3</a>,
            <a href="http://docs.getpelican.com/" target="_blank">Pelican</a>,
            <a href="http://getbootstrap.com" target="_blank">Bootstrap</a>              <p><small>  <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="/static/images/by-sa-80x15.png" /></a>
    &quot;<span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">Huang Huang 的博客</span>&quot; by <a xmlns:cc="http://creativecommons.org/ns#" href="https://mozillazg.com" property="cc:attributionName" rel="cc:attributionURL">mozillazg</a> is
  licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>, except where indicated otherwise.
</small></p>
         </div>
         <div class="col-xs-2"><p class="pull-right"><i class="fa fa-arrow-up"></i> <a href="#">Back to top</a></p></div>
      </div>
   </div>
</footer>
<script src="https://mozillazg.com/theme/js/jquery.min.js"></script>

<!-- Include all compiled plugins (below), or include individual files as needed -->
<script src="https://mozillazg.com/theme/js/bootstrap.min.js"></script>

<!-- Enable responsive features in IE8 with Respond.js (https://github.com/scottjehl/Respond) -->
<script src="https://mozillazg.com/theme/js/respond.min.js"></script>

    <!-- Disqus -->
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'my-github-blog'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function () {
            var s = document.createElement('script');
            s.async = true;
            s.type = 'text/javascript';
            s.src = '//' + disqus_shortname + '.disqus.com/count.js';
            (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
        }());
    </script>
    <!-- End Disqus Code -->
    <!-- Google Analytics Universal -->
    <script type="text/javascript">
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

        ga('create', 'UA-77172981-1', 'auto');
        ga('send', 'pageview');
    </script>
    <!-- End Google Analytics Universal Code -->


<script src="https://mozillazg.com/theme/js/vdonate.js"></script>
<script>
$(document).ready(function(){
new vdonate({
  title: '如果觉得我的文章对您有用，请随意打赏。您的支持将鼓励我继续创作!', // 可选参数，打赏标题
  btnText: '打赏支持', // 可选参数，打赏按钮文字
  el: document.getElementById('donate'), // 必选参数，打赏按钮的容器
  wechatImage: 'https://mozillazg.com/static/wechat_donate.png', // 必选参数，微信收款二维码
  alipayImage: 'https://mozillazg.com/static/alipay_donate.jpg' // 必选参数，支付宝收款二维码
});
});
</script>

<script src="https://mozillazg.com/theme/js/tocbot.min.js"></script>
<script>
$(document).ready(function(){
  tocbot.init({
    // Where to render the table of contents.
    tocSelector: '.js-toc',
    // Where to grab the headings to build the table of contents.
    contentSelector: '.entry-content',
    // Which headings to grab inside of the contentSelector element.
    headingSelector: 'h2, h3, h4,h5',
    // Headings that match the ignoreSelector will be skipped.
    ignoreSelector: '.js-toc-ignore',
    // Main class to add to links.
    linkClass: 'toc-link',
    // Extra classes to add to links.
    extraLinkClasses: '',
    // Class to add to active links,
    // the link corresponding to the top most heading on the page.
    activeLinkClass: 'is-active-link',
    // Main class to add to lists.
    listClass: 'toc-list',
    // Extra classes to add to lists.
    extraListClasses: '',
    // Class that gets added when a list should be collapsed.
    isCollapsedClass: 'is-collapsed',
    // Class that gets added when a list should be able
    // to be collapsed but isn't necessarily collpased.
    collapsibleClass: 'is-collapsible',
    // Class to add to list items.
    listItemClass: 'toc-list-item',
    // How many heading levels should not be collpased.
    // For example, number 6 will show everything since
    // there are only 6 heading levels and number 0 will collpase them all.
    // The sections that are hidden will open
    // and close as you scroll to headings within them.
    collapseDepth: 6,
    // Smooth scrolling enabled.
    smoothScroll: true,
    // Smooth scroll duration.
    smoothScrollDuration: 420,
    // Callback for scroll end (requires: smoothScroll).
    scrollEndCallback: function (e) {},
    // Headings offset between the headings and the top of the document (this is meant for minor adjustments).
    headingsOffset: 0,
    // Timeout between events firing to make sure it's
    // not too rapid (for performance reasons).
    throttleTimeout: 50,
    // Element to add the positionFixedClass to.
    positionFixedSelector: null,
    // Fixed position class to add to make sidebar fixed after scrolling
    // down past the fixedSidebarOffset.
    positionFixedClass: 'is-position-fixed',
    // fixedSidebarOffset can be any number but by default is set
    // to auto which sets the fixedSidebarOffset to the sidebar
    // element's offsetTop from the top of the document on init.
    fixedSidebarOffset: 'auto',
    // includeHtml can be set to true to include the HTML markup from the
    // heading node instead of just including the textContent.
    includeHtml: false
  });
});
</script>
</body>
</html>